File: algorithm_data_structure/leetcode/46.permutations.py
#
# @lc app=leetcode id=46 lang=python3
#
# [46] Permutations
#

# @lc code=start
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # The backtracking is used to enumerate all situations
        res = []
        if len(nums) == 0:
            return res
        self.get_permute(res, nums, 0)
        return res

    def get_permute(self, res, nums, index):
        if index == len(nums):
            res.append(list(nums))
            return

        for i in range(index, len(nums)):
            # Take action!
            nums[i], nums[index] = nums[index], nums[i]
            # Backtracking
            self.get_permute(res, nums, index + 1)
            # Recover action
            nums[i], nums[index] = nums[index], nums[i]


# @lc code=end

File: algorithm_data_structure/leetcode/46.permutations.cpp
/*
 * @lc app=leetcode id=46 lang=cpp
 *
 * [46] Permutations
 */

// @lc code=start

#include <algorithm>
#include <vector>

using namespace std;

class Solution
{
public:
    void backtracking(vector<int> &nums, int level, vector<vector<int>> &ans)
    {
        // Keep swap the level and level + 1 elements. So the maxmial allowed level is nums.size() - 1
        if (level == nums.size() - 1)
        {
            ans.push_back(nums);
            return;
        }

        for (int i = level; i < nums.size(); i++)
        {
            swap(nums[i], nums[level]);
            backtracking(nums, level + 1, ans); // Recursively iterate over the next position
            swap(nums[i], nums[level]);         // Backtrack to the original state
        }
    }

    vector<vector<int>> permute(vector<int> &nums)
    {
        vector<vector<int>> ans;
        backtracking(nums, 0, ans);
        return ans;
    }
};
// @lc code=end
